spec for codestore error (#393)
privacy-scaling-explorations/zkevm-specs: specs/error_state/ErrorCodeStore.md

# ErrorCodeStore state

## Procedure
`ErrorCodeStore` is for two original code store related errors: `CodeStoreOutOfGas` and 
`MaxCodeSizeExceeded`.
this type of error only occurs when executing create(create,create2) op code or tx deploy 
transaction(tx.to = null).

### EVM behavior

> when in create kind transaction, init bytecodes will run and current call context is creating.
> the final bytecodes to store for new contract is `RETURN` opcode of init codes result. `RETURN` opcode returns
> memory [`offset`...`offset` + `length`] content as bytecodes to store into state db.
> for returned bytecodes, store them cost additional gas.   

Reviewer: ```suggestion
When handling a CREATE-kind transaction, Initial bytecode opcodes will run and current call context is created.
The final bytecode opcodes to store for new contract is the `RETURN` opcode of init codes result.

`RETURN` opcode returns memory [`offset`...`offset` + `length`] content as bytecodes to store into state db.
for returned bytecodes, store them cost additional gas.   
```

`let CODE_DEPOSIT_BYTE_COST = 200
code_store_cost = CODE_DEPOSIT_BYTE_COST * bytecodes' length
`  

if `code_store_cost` > gas left, it is `CodeStoreOutOfGas` case.  
if returned bytecodes' length > `MAXCODESIZE` allowed in evm, it is 
`MaxCodeSizeExceeded` case.  

> in circuit buss mapping side, check these two code store errors in [https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/main/bus-mapping/src/circuit_input_builder/input_state_ref.rs#L1148&L1155]

Reviewer: ```suggestion
In circuit bus mapping side, check these two code store errors in [here](https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/main/bus-mapping/src/circuit_input_builder/input_state_ref.rs#L1148&L1155)
```

> when executing op code is `RETURN` and call context is creating(`call.is_create == true`) at the meanwhile.  

Reviewer: ```suggestion
when executing opcode is `RETURN` and call context is creating(`call.is_create == true`) meanwhile.  
```

> even though errors occur in `create` kind op codes, it is special not checking error 
> in executing op code `create` directly.  
> circuit implementation take similar strategy, not constrain error directly in create op codes, but 
> in `RETURN` step context. following this way it easy to get the key property state `length` and construct constraint
> against it.
> 
> overall as following:  
> Pop EVM word `offset` and `length` from the stack, 
> then go to `ErrorCodeStore` state when call context is creating & 
> one of the followings occurs:

Reviewer: ```suggestion
Even though errors occur in `CREATE` kind opcodes, it is special not checking error 
in executing opcode `CREATE` directly.  
Circuit implementation takes similar strategy, not constrain error directly in CREATE opcodes, but 
in `RETURN` step context. Following this way it easy to get the key property state `length` and construct constraints against it.

Overall it looks like the following:  
- Pop EVM word `offset` and `length` from the stack, 
- Go to `ErrorCodeStore` state when call context is being created & 
select which one of the followings occurs:
```

> -  storing `length` of bytecodes out of gas
> -  `length` of bytecodes exceeds `MAXCODESIZE`

Reviewer: ```suggestion
1.  Storing `length` of bytecodes  runs out of gas.
2.  `length` of bytecodes exceeds `MAXCODESIZE`.
```

### Constraints
1. `code_store_cost` > gas_left or `length` > `MAXCODESIZE`

> 2. Current call must be failed.

Reviewer: Current call should not be STATICALL. Do we need to highlight it here?

Author: @CPerezz  you are right ! will add this constraint.

Reviewer: Aside from this one, I wonder if we're missing something else. Specially regarding call-depth or similar.

Author: I realize that each constraint you mention has its error gadget handling. like if create in `STATICALL` ,  ErrorWriteProtection gadget will handle it.  if call-depth exceeds, it will go to ErrorDepth gadget side. so I don't think need check each one. otherwise we will ensure all other errors of create(ErrDepth, ErrInsufficientBalance, ErrContractAddressCollision , ErrInvalidCode etc.) not happens.

Author: added non static call constraint in [cbb1600](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/393/commits/cbb1600bf1cdb7b5f6d28b548dd2f49ed3e56185)
